Analyzing car sales data from "eBay Kleinanzeigen"
Modified dataset from "Kaggle". Consisting of 50,000 data points.


    We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
    We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

https://www.kaggle.com/orgesleka/used-cars-database/data

In [1]:
import pandas as pd
import numpy as np

autos=pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changed the column names according to python standards
Changed CamelCase to snake_case

In [6]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

drop_candidates:
seller does not offer additional information, since almost all are "privat"
its the same with "offer type"

convert:
price needs to be converted to int
odometer should be converted to a numerical dtype

examine:
registration year, month, ad_created, last_seen need to be further examined: It may be better to convert them to a datetime object
whats wrong with nr. of pictures? --> further examination

In [7]:
#examination
autos["registration_year"].unique()
#some are just false. How frequent are those?

print(autos.loc[autos["registration_year"]<1900,"registration_year"].value_counts().sum())
print(autos.loc[autos["registration_year"]>2018,"registration_year"].value_counts().sum())

6
21


At least 27 data points have a wrong registration year.
Because its really just the year that is stored you can leave it as int

In [8]:
autos["registration_month"].unique()
#0 is impossible, how frequent are those values

print(autos.loc[autos["registration_month"]==0,"registration_month"].value_counts().sum())

5075


At least 5075 data points in this column represent missing data and should be converted to np.nan

It might be an option to merge registration year and month to one datetime object for further analysis (if needed).

In [9]:
autos["ad_created"].unique()

type(autos.loc[0,"ad_created"])

str

what appears to be a datetime object is in reality a string object --> Convert

hours, minuted, seconds should be scrapped from the datetime object because they provide no information at all.

In [10]:
autos["last_seen"].unique()

type(autos.loc[0,"last_seen"])

str

last_seen should also be converted to datetime.
Plausability analysis can also be done after the conversion to datetime.

In [11]:
#wtf is wrong with nr of pictures?
autos["nr_of_pictures"].unique()

autos["nr_of_pictures"].head()

0    0
1    0
2    0
3    0
4    0
Name: nr_of_pictures, dtype: int64

this column should be dropped entirely. It provides no information.
Either none of the ads had a picture in it(not very plausible) or there was some kind of error in the data collection process(plausible).

Converting price to a numeric dtype

In [12]:
print(autos["price"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [13]:
#I have to get rid of the "$" string as well as the comma.

autos["price"]=(autos["price"]
               .str.replace("$","")
               .str.replace(",","")
               .str.strip()
               .astype(int)
               )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

Converting odometer to a numeric dtype

In [14]:
print(autos["odometer"].unique())

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [15]:
#I have to get rid of the comma and the "km" string

autos["odometer"]=(autos["odometer"]
                   .str.replace("km","")
                   .str.replace(",","")
                   .str.strip()
                   .astype(int)
                  )
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [16]:
#I have to rename the column name in order to make up for the information loss by deleting km

autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Removing outliers from the columns "odometer_km" and "price"

In [17]:
autos["odometer_km"].unique().shape

(13,)

In [18]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

The values seem reasonable. No outliers detected.

Analyzing the price column

In [20]:
autos["price"].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [21]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Min is 0
Max is way too high (100.000.000)

In [22]:
autos["price"].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [23]:
autos["price"].value_counts().sort_index(ascending=True).head(100)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
       ... 
215       2
217       1
219       1
220      33
222      12
225       8
230      12
235       2
238       1
240       3
248       1
249      13
250     291
251       1
255       1
260       5
269       1
270       6
275       7
277       1
280      30
285       1
290      19
295       1
299      56
300     384
310       1
320      12
325       5
329       2
Name: price, Length: 100, dtype: int64

I choose to remove all outliers between the price of under 50 and above 350000

In [24]:
autos=autos[autos["price"].between(50,350000)]

In [25]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48347 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          48347 non-null object
name                  48347 non-null object
seller                48347 non-null object
offer_type            48347 non-null object
price                 48347 non-null int64
abtest                48347 non-null object
vehicle_type          43874 non-null object
registration_year     48347 non-null int64
gearbox               46096 non-null object
power_ps              48347 non-null int64
model                 45931 non-null object
odometer_km           48347 non-null int64
registration_month    48347 non-null int64
fuel_type             44418 non-null object
brand                 48347 non-null object
unrepaired_damage     39397 non-null object
ad_created            48347 non-null object
nr_of_pictures        48347 non-null int64
postal_code           48347 non-null int64
last_seen             48347 non-null object
dtypes: int64(7), 

In [26]:
autos[ "price"].describe()

count     48347.000000
mean       5915.456099
std        9071.625794
min          50.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

Analyzing the distribution of the values in date_crawled, ad_created and last_seen (all strings) as percentages.

In [27]:
#date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036031
2016-03-08    0.033260
2016-03-09    0.033053
2016-03-10    0.032225
2016-03-11    0.032556
2016-03-12    0.036983
2016-03-13    0.015658
2016-03-14    0.036672
2016-03-15    0.034294
2016-03-16    0.029495
2016-03-17    0.031584
2016-03-18    0.012907
2016-03-19    0.034749
2016-03-20    0.037789
2016-03-21    0.037293
2016-03-22    0.032949
2016-03-23    0.032267
2016-03-24    0.029433
2016-03-25    0.031501
2016-03-26    0.032267
2016-03-27    0.031088
2016-03-28    0.034894
2016-03-29    0.034108
2016-03-30    0.033715
2016-03-31    0.031832
2016-04-01    0.033735
2016-04-02    0.035535
2016-04-03    0.038575
2016-04-04    0.036528
2016-04-05    0.013051
2016-04-06    0.003165
2016-04-07    0.001386
Name: date_crawled, dtype: float64

The first crawling took place on march the 7nd, 2016

In [28]:
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=False)

2016-04-07    0.001386
2016-04-06    0.003165
2016-04-05    0.013051
2016-04-04    0.036528
2016-04-03    0.038575
2016-04-02    0.035535
2016-04-01    0.033735
2016-03-31    0.031832
2016-03-30    0.033715
2016-03-29    0.034108
2016-03-28    0.034894
2016-03-27    0.031088
2016-03-26    0.032267
2016-03-25    0.031501
2016-03-24    0.029433
2016-03-23    0.032267
2016-03-22    0.032949
2016-03-21    0.037293
2016-03-20    0.037789
2016-03-19    0.034749
2016-03-18    0.012907
2016-03-17    0.031584
2016-03-16    0.029495
2016-03-15    0.034294
2016-03-14    0.036672
2016-03-13    0.015658
2016-03-12    0.036983
2016-03-11    0.032556
2016-03-10    0.032225
2016-03-09    0.033053
2016-03-08    0.033260
2016-03-07    0.036031
2016-03-06    0.014065
2016-03-05    0.025358
Name: date_crawled, dtype: float64

In [29]:
autos["date_crawled"].str[:10].describe()

count          48347
unique            34
top       2016-04-03
freq            1865
Name: date_crawled, dtype: object

The first crawling took place on march the 7nd, 2016.
The last crawling took place on April the 7th, 2016.
The most crawling took place on April the 3rd, 2016. (3,86%)

In [30]:
#ad_created

autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=False)

2016-04-07    0.001241
2016-04-06    0.003247
2016-04-05    0.011769
2016-04-04    0.036900
2016-04-03    0.038824
2016-04-02    0.035225
2016-04-01    0.033715
2016-03-31    0.031874
2016-03-30    0.033528
2016-03-29    0.034066
2016-03-28    0.034997
2016-03-27    0.031005
2016-03-26    0.032329
2016-03-25    0.031626
2016-03-24    0.029371
2016-03-23    0.032101
2016-03-22    0.032763
2016-03-21    0.037520
2016-03-20    0.037851
2016-03-19    0.033632
2016-03-18    0.013589
2016-03-17    0.031253
2016-03-16    0.029992
2016-03-15    0.034025
2016-03-14    0.035307
2016-03-13    0.017023
2016-03-12    0.036796
2016-03-11    0.032887
2016-03-10    0.031936
2016-03-09    0.033135
                ...   
2016-02-21    0.000062
2016-02-20    0.000041
2016-02-19    0.000062
2016-02-18    0.000041
2016-02-17    0.000021
2016-02-16    0.000021
2016-02-14    0.000041
2016-02-12    0.000041
2016-02-11    0.000021
2016-02-09    0.000021
2016-02-08    0.000021
2016-02-07    0.000021
2016-02-05 

In [31]:
autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033135
2016-03-10    0.031936
2016-03-11    0.032887
2016-03-12    0.036796
2016-03-13    0.017023
2016-03-14    0.035307
2016-03-15    0.034025
2016-03-16    0.029992
2016-03-17    0.031253
2016-03-18    0.013589
2016-03-19    0.033632
2016-03-20    0.037851
2016-03-21 

In [32]:
autos["ad_created"].str[:10].describe()

count          48347
unique            76
top       2016-04-03
freq            1877
Name: ad_created, dtype: object

The dataset covers ads that were created between November the 6th, 2015 and April the 7th, 2016.
Most ads were created on April the 3rd 2016. (3,88%)

In [33]:
#Last_seen
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=False)

2016-04-07    0.132087
2016-04-06    0.221627
2016-04-05    0.124972
2016-04-04    0.024510
2016-04-03    0.025172
2016-04-02    0.024883
2016-04-01    0.022876
2016-03-31    0.023828
2016-03-30    0.024738
2016-03-29    0.022318
2016-03-28    0.020849
2016-03-27    0.015575
2016-03-26    0.016775
2016-03-25    0.019174
2016-03-24    0.019774
2016-03-23    0.018595
2016-03-22    0.021408
2016-03-21    0.020601
2016-03-20    0.020622
2016-03-19    0.015802
2016-03-18    0.007322
2016-03-17    0.028089
2016-03-16    0.016464
2016-03-15    0.015885
2016-03-14    0.012638
2016-03-13    0.008873
2016-03-12    0.023745
2016-03-11    0.012390
2016-03-10    0.010611
2016-03-09    0.009639
2016-03-08    0.007343
2016-03-07    0.005419
2016-03-06    0.004323
2016-03-05    0.001076
Name: last_seen, dtype: float64

In [34]:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.001076
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007343
2016-03-09    0.009639
2016-03-10    0.010611
2016-03-11    0.012390
2016-03-12    0.023745
2016-03-13    0.008873
2016-03-14    0.012638
2016-03-15    0.015885
2016-03-16    0.016464
2016-03-17    0.028089
2016-03-18    0.007322
2016-03-19    0.015802
2016-03-20    0.020622
2016-03-21    0.020601
2016-03-22    0.021408
2016-03-23    0.018595
2016-03-24    0.019774
2016-03-25    0.019174
2016-03-26    0.016775
2016-03-27    0.015575
2016-03-28    0.020849
2016-03-29    0.022318
2016-03-30    0.024738
2016-03-31    0.023828
2016-04-01    0.022876
2016-04-02    0.024883
2016-04-03    0.025172
2016-04-04    0.024510
2016-04-05    0.124972
2016-04-06    0.221627
2016-04-07    0.132087
Name: last_seen, dtype: float64

In [35]:
autos["last_seen"].str[:10].describe()

count          48347
unique            34
top       2016-04-06
freq           10715
Name: last_seen, dtype: object

The ads were last seen between March the 5th 2016 and April the 7th 2016.
The most common date was April the 6th, 2016 (22,16%).

Understand the distribution of registration_year

In [36]:
type(autos.loc[0,"registration_year"])

numpy.int64

In [37]:
autos["registration_year"].describe()

count    48347.000000
mean      2004.718845
std         87.786604
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Arithmetisches Mittel und Median stimmen zum großen Teil überein --> Wenig Einfluss durch Extremwerte --> Verteilung interpretierbar.
Min und Max sind nicht interpretierbar, da offentsichtlich unplausibel.
Das erste Viertel ist in den vor der Jahrtausendwende registriert worden.
Die oberen 25% nach 2008.
Die Hälfte der Autos hat ihre Erstzulassung nach 2004 gehabt, die andere Hälfte davor.

Further examination to decide wether it is okay to remove all the rows that have a registration year before the 1900s and above 2016

In [38]:
autos["registration_year"].value_counts(dropna=False).sort_index(ascending=True)

1000       1
1001       1
1111       1
1800       2
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      22
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3120
2001    2632
2002    2478
2003    2695
2004    2699
2005    2921
2006    2670
2007    2274
2008    2212
2009    2081
2010    1589
2011    1620
2012    1309
2013     801
2014     662
2015     384
2016    1207
2017    1385
2018     468
2019       2
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

the highest value will be 2016 because the crawled data is from 2016.
the lowest value will be 1910 because this is the lowest plausible value for a car maufacturing date.

In [39]:
autos=autos.loc[autos["registration_year"].between(1910,2016)]

autos["registration_year"].value_counts(dropna=False).sort_index()

1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      22
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
        ... 
1987      72
1988     134
1989     171
1990     336
1991     339
1992     368
1993     424
1994     629
1995    1205
1996    1364
1997    1940
1998    2351
1999    2884
2000    3120
2001    2632
2002    2478
2003    2695
2004    2699
2005    2921
2006    2670
2007    2274
2008    2212
2009    2081
2010    1589
2011    1620
2012    1309
2013     801
2014     662
2015     384
2016    1207
Name: registration_year, Length: 78, dtype: int64

In [40]:
autos["registration_year"].value_counts(normalize=True)

2000    0.067136
2005    0.062854
1999    0.062058
2004    0.058077
2003    0.057991
2006    0.057453
2001    0.056635
2002    0.053321
1998    0.050589
2007    0.048932
2008    0.047598
2009    0.044779
1997    0.041745
2011    0.034859
2010    0.034192
1996    0.029350
2012    0.028167
2016    0.025972
1995    0.025929
2013    0.017236
2014    0.014245
1994    0.013535
1993    0.009124
2015    0.008263
1992    0.007919
1991    0.007295
1990    0.007230
1989    0.003680
1988    0.002883
1985    0.002023
          ...   
1966    0.000473
1976    0.000452
1969    0.000409
1975    0.000387
1965    0.000366
1964    0.000258
1963    0.000172
1961    0.000129
1959    0.000129
1962    0.000086
1956    0.000086
1937    0.000086
1958    0.000086
1955    0.000043
1954    0.000043
1957    0.000043
1951    0.000043
1910    0.000043
1934    0.000043
1941    0.000043
1927    0.000022
1929    0.000022
1950    0.000022
1931    0.000022
1948    0.000022
1938    0.000022
1939    0.000022
1953    0.0000

Most of the cars were first registred in the 2000s, the 90s or late 80s.
oldtimers are the fewest.

Explore the unique values in the brand column, and decide on which brands you want to aggregate by. 

In [41]:
autos["brand"].unique().shape

(40,)

In [42]:
autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(10)

volkswagen       0.211327
bmw              0.109956
opel             0.107310
mercedes_benz    0.096508
audi             0.086609
ford             0.069976
renault          0.047189
peugeot          0.029910
fiat             0.025714
seat             0.018247
Name: brand, dtype: float64

In [43]:
autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(10).sum()

0.8027456802874788

I have chosen to look at the top 10 of car manufacturers, because they make up over 80% of the used market.

The next step is to calculate the mean prices for each brand

In [44]:
top_10_brands=autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(10).index
print(top_10_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [45]:
top_10_mean_price_by_brand={}

for brand in list(top_10_brands):
    top_10_mean_price_by_brand[brand]=autos.loc[autos["brand"]==brand,"price"].mean()
    
print(top_10_mean_price_by_brand)

{'renault': 2483.889192886457, 'opel': 2996.0850210547424, 'seat': 4423.116745283019, 'volkswagen': 5424.9185419000105, 'mercedes_benz': 8663.059308807135, 'fiat': 2818.456066945607, 'peugeot': 3100.6928057553955, 'bmw': 8376.805088062622, 'ford': 3762.1165436654364, 'audi': 9373.771428571428}


In [46]:
#because dicts are orderless, I will transform into a series to perform further analysis.

top_10_mean_price_by_brand_df=pd.DataFrame.from_dict(top_10_mean_price_by_brand,orient="index",)
print(top_10_mean_price_by_brand_df.sort_values(0,ascending=False))

                         0
audi           9373.771429
mercedes_benz  8663.059309
bmw            8376.805088
volkswagen     5424.918542
seat           4423.116745
ford           3762.116544
peugeot        3100.692806
opel           2996.085021
fiat           2818.456067
renault        2483.889193


On average Audi ist the most expensive brand for used cars, followed by Mercedes and BMW. The gap between those brands is more or less the same.

The next segment is about 3000 Euros away for the first. It consists of Volkswagen, Seat, Ford and Peugeot. Those have an average price gap of about 700 Euros.

The next segment consist of the sub 3000 Euros on average car brands Opel, Fiat and Renault. The average price gaps are between 100 and 350 Euros.

Mean mileage by brand

In [47]:
mean_milage_brand={}

for brand in top_10_brands:
    mean_milage_brand[brand]=autos.loc[autos["brand"]==brand,"odometer_km"].mean()
    
mean_milage_brand_series=pd.Series(mean_milage_brand)
mean_price_brand_series=pd.Series(top_10_mean_price_by_brand)

means_top_10=pd.DataFrame(mean_milage_brand_series,columns=["mean_milage"])
means_top_10

mean_milage
audi           129254.658385
bmw            132651.663405
fiat           117066.945607
ford           124295.817958
mercedes_benz  130923.076923
opel           129440.545418
peugeot        127122.302158
renault        128144.094847
seat           121303.066038
volkswagen     128771.000916

In [48]:
means_top_10["mean_price"]=mean_price_brand_series
means_top_10

mean_milage   mean_price
audi           129254.658385  9373.771429
bmw            132651.663405  8376.805088
fiat           117066.945607  2818.456067
ford           124295.817958  3762.116544
mercedes_benz  130923.076923  8663.059309
opel           129440.545418  2996.085021
peugeot        127122.302158  3100.692806
renault        128144.094847  2483.889193
seat           121303.066038  4423.116745
volkswagen     128771.000916  5424.918542

In [49]:
print(means_top_10.sort_values(by="mean_price",axis=0,ascending=False))

                 mean_milage   mean_price
audi           129254.658385  9373.771429
mercedes_benz  130923.076923  8663.059309
bmw            132651.663405  8376.805088
volkswagen     128771.000916  5424.918542
seat           121303.066038  4423.116745
ford           124295.817958  3762.116544
peugeot        127122.302158  3100.692806
opel           129440.545418  2996.085021
fiat           117066.945607  2818.456067
renault        128144.094847  2483.889193


There seems to be a very small correlation at best, if not none.

Further analysis:
1. Translate categorial data to english

In [50]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [51]:
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [52]:
vt_map={"bus":"bus","limousine":"limousine","kleinwagen":"compact_car","kombi":"wagon","coupe":"coupe","suv":"suv","cabrio":"convertible","andere":"other"}

print(autos["vehicle_type"].value_counts())
print(autos["vehicle_type"].isnull().sum())
autos["vehicle_type"]=autos["vehicle_type"].map(vt_map)
print(autos["vehicle_type"].value_counts())
print(autos["vehicle_type"].isnull().sum())

limousine     12567
kleinwagen    10555
kombi          8911
bus            4024
cabrio         3008
coupe          2456
suv            1963
andere          388
Name: vehicle_type, dtype: int64
2601
limousine      12567
compact_car    10555
wagon           8911
bus             4024
convertible     3008
coupe           2456
suv             1963
other            388
Name: vehicle_type, dtype: int64
2601


Convert the dates to be uniform numeric data

In [53]:
autos[['date_crawled','registration_year','last_seen','ad_created']]

date_crawled  registration_year            last_seen  \
0      2016-03-26 17:47:46               2004  2016-04-06 06:45:54   
1      2016-04-04 13:38:56               1997  2016-04-06 14:45:08   
2      2016-03-26 18:57:24               2009  2016-04-06 20:15:37   
3      2016-03-12 16:58:10               2007  2016-03-15 03:16:28   
4      2016-04-01 14:38:50               2003  2016-04-01 14:38:50   
5      2016-03-21 13:47:45               2006  2016-04-06 09:45:21   
6      2016-03-20 17:55:21               1995  2016-03-23 02:48:59   
7      2016-03-16 18:55:19               1998  2016-04-07 03:17:32   
8      2016-03-22 16:51:34               2000  2016-03-26 18:18:10   
9      2016-03-16 13:47:02               1997  2016-04-06 10:46:35   
11     2016-03-16 18:45:34               2000  2016-03-16 18:45:34   
12     2016-03-31 19:48:22               2010  2016-04-06 14:17:52   
13     2016-03-23 10:48:32               1999  2016-04-01 14:17:13   
14     2016-03-23 11:50:46               2007  2016-04-01 15:46:47   
15     2016-04-01 12:06:20               1982  2016-04-02 21:10:48   
16     2016-03-16 14:59:02               1999  2016-03-18 05:29:37   
17     2016-03-29 11:46:22               1990  2016-04-05 20:46:26   
18     2016-03-26 19:57:44               1995  2016-04-02 12:16:41   
19     2016-03-17 13:36:21               2004  2016-03-17 14:45:58   
20     2016-03-05 19:57:31               2003  2016-03-07 05:46:46   
21     2016-03-06 19:07:10               2004  2016-04-05 23:46:19   
22     2016-03-28 20:50:54               2015  2016-04-01 06:45:30   
23     2016-03-10 19:55:34               2010  2016-03-17 08:45:17   
24     2016-04-03 11:57:02               2014  2016-04-07 13:16:50   
25     2016-03-21 21:56:18               1996  2016-04-01 05:16:49   
26     2016-04-03 22:46:28               1992  2016-04-05 23:46:48   
28     2016-03-19 21:56:19               2007  2016-04-07 14:58:48   
29     2016-04-02 12:45:44               2004  2016-04-02 12:45:44   
30     2016-03-14 11:47:31               2002  2016-03-14 11:47:31   
31     2016-03-14 16:53:09               2002  2016-03-16 07:19:04   
...                    ...                ...                  ...   
49968  2016-04-01 17:49:15               1986  2016-04-05 13:16:35   
49969  2016-03-17 18:49:02               2005  2016-03-25 23:18:15   
49970  2016-03-21 22:47:37               2010  2016-04-07 04:17:34   
49971  2016-03-29 14:54:12               2001  2016-03-29 20:41:51   
49972  2016-03-26 22:25:23               2004  2016-03-28 11:28:18   
49973  2016-03-27 05:32:39               2004  2016-03-27 08:25:24   
49975  2016-03-27 20:51:39               2012  2016-04-05 19:45:34   
49976  2016-03-19 18:56:05               1992  2016-04-07 06:16:44   
49977  2016-03-31 18:37:18               2003  2016-04-06 12:16:11   
49978  2016-04-04 10:37:14               1996  2016-04-06 12:44:20   
49979  2016-03-20 18:38:40               2011  2016-04-07 01:45:12   
49980  2016-03-12 10:55:54               1995  2016-04-06 17:16:49   
49981  2016-03-15 09:38:21               1998  2016-04-05 17:21:46   
49982  2016-03-29 18:51:08               2004  2016-03-29 18:51:08   
49983  2016-03-06 12:43:04               1999  2016-03-09 06:16:08   
49985  2016-04-02 16:38:23               1995  2016-04-06 15:17:52   
49986  2016-04-04 20:46:02               2010  2016-04-06 23:16:00   
49987  2016-03-22 20:47:27               2013  2016-03-26 22:46:06   
49988  2016-03-28 19:49:51               2001  2016-04-07 02:17:40   
49989  2016-03-11 19:50:37               1997  2016-03-12 10:17:55   
49990  2016-03-21 19:54:19               2012  2016-04-06 22:46:57   
49991  2016-03-06 15:25:19               2016  2016-03-06 18:24:19   
49992  2016-03-10 19:37:38               2009  2016-03-13 01:44:51   
49993  2016-03-15 18:47:35               1997  2016-04-06 19:46:53   
49994  2016-03-22 17:36:42               2001  2016-04-06 08:16:39   
49995  2016-03-27 14:38:

In [54]:
#Write a function for the three date columns

def clean_date(column):
    clean_c=column.str.split(n=1,expand=True)
    clean_c.drop(1,axis=1,inplace=True)
    clean_c=(clean_c[0]
            .str.replace("-","")
            .str.strip()
            .astype(int)
            )
    return clean_c

autos["date_crawled"]=clean_date(autos["date_crawled"])
autos["date_crawled"]

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
5        20160321
6        20160320
7        20160316
8        20160322
9        20160316
11       20160316
12       20160331
13       20160323
14       20160323
15       20160401
16       20160316
17       20160329
18       20160326
19       20160317
20       20160305
21       20160306
22       20160328
23       20160310
24       20160403
25       20160321
26       20160403
28       20160319
29       20160402
30       20160314
31       20160314
           ...   
49968    20160401
49969    20160317
49970    20160321
49971    20160329
49972    20160326
49973    20160327
49975    20160327
49976    20160319
49977    20160331
49978    20160404
49979    20160320
49980    20160312
49981    20160315
49982    20160329
49983    20160306
49985    20160402
49986    20160404
49987    20160322
49988    20160328
49989    20160311
49990    20160321
49991    20160306
49992    20160310
49993    20160315
49994    2

In [55]:
autos["last_seen"]=clean_date(autos["last_seen"])
autos["last_seen"]

0        20160406
1        20160406
2        20160406
3        20160315
4        20160401
5        20160406
6        20160323
7        20160407
8        20160326
9        20160406
11       20160316
12       20160406
13       20160401
14       20160401
15       20160402
16       20160318
17       20160405
18       20160402
19       20160317
20       20160307
21       20160405
22       20160401
23       20160317
24       20160407
25       20160401
26       20160405
28       20160407
29       20160402
30       20160314
31       20160316
           ...   
49968    20160405
49969    20160325
49970    20160407
49971    20160329
49972    20160328
49973    20160327
49975    20160405
49976    20160407
49977    20160406
49978    20160406
49979    20160407
49980    20160406
49981    20160405
49982    20160329
49983    20160309
49985    20160406
49986    20160406
49987    20160326
49988    20160407
49989    20160312
49990    20160406
49991    20160306
49992    20160313
49993    20160406
49994    2

In [56]:
autos["ad_created"]=clean_date(autos["ad_created"])
autos["ad_created"]

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
5        20160321
6        20160320
7        20160316
8        20160322
9        20160316
11       20160316
12       20160331
13       20160323
14       20160323
15       20160401
16       20160316
17       20160329
18       20160326
19       20160317
20       20160305
21       20160306
22       20160328
23       20160310
24       20160403
25       20160321
26       20160403
28       20160319
29       20160402
30       20160314
31       20160314
           ...   
49968    20160401
49969    20160317
49970    20160321
49971    20160329
49972    20160326
49973    20160327
49975    20160327
49976    20160319
49977    20160331
49978    20160404
49979    20160320
49980    20160312
49981    20160315
49982    20160329
49983    20160306
49985    20160402
49986    20160404
49987    20160322
49988    20160328
49989    20160311
49990    20160321
49991    20160306
49992    20160310
49993    20160315
49994    2

In [57]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [58]:
autos["name"].unique()

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United', ...,
       'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon',
       'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!',
       'Fiat_500_C_1.2_Dualogic_Lounge'], dtype=object)

In [59]:
autos["name"].unique().shape

(35613,)

Möglichkeiten der Auslagerung wären Hubraum, Navi, usw.
Ich versuche es beispielhaft dem Navi

In [60]:
navi_regex="navi"

navi=autos.loc[autos["name"].str.contains(navi_regex,case=False),"name"]
navi

34                                 Bmw_530d__navi_tuv03/18
159      Mini_Cooper_S__Leder__Xenon__Navi__Harman_Kard...
194        BMW_X3_xDrive20d_Aut./X_Line/Navi/Xenon/PDC/AHK
200      Corvette_C6_Grand_Sport_Cabrio_Schalter_EU_Mod...
212            Porsche_Boxster_Tiptronic_S_unfallfrei_Navi
321             Audi_A1_1.4_TFSI_S_tronic_S_line_Navi_Bose
332        Suzuki_Swift_Comfort+_Navi_Klima_Keyless_go_LPG
409          Opel_Corsa_1.2_16V_Navi100_AHK_Tempomat_EURO4
412      Volkswagen_Golf_2.0_TDI_DPF_Team___Navi__Park_...
420      Audi_Audi_A6_Avant_3.0_TDI_S_Line_Navi_DSG_Xen...
462      Porsche_Boxster_S__Volleder__Navi__appr._Garantie
482      BMW_520d_Touring__Navi_Prof.__AHK__Standhz._Le...
504      Volkswagen_Golf_2.0_TDI_DSG_GT_Leder_Xenon_Nav...
515      Volkswagen_Multivan_DPF_Atlantis_2.Hand_Navi_T...
566      Mercedes_Benz_BlueEFFICIENCY_7G_TRONIC_Avantga...
674      Audi_A4_Avant_2.0_TDI_multitronic_Navi_Xenon_8...
681        Toyota_IQ_1.4_D_4D_+_Leder_Navi__8_fach_berei

In [61]:
autos["nav"]=navi
autos.loc[autos["nav"].isnull()==0,"nav"]="Yes"
autos.loc[autos["nav"].isnull(),"nav"]="Unknown"
autos["nav"].value_counts()

Unknown    45090
Yes         1383
Name: nav, dtype: int64

What are the most common brand/models combinations?

In [62]:
autos["model"].unique()

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'wrangler',
       'a6', 'transporter', 'astra', 'v40', 'ibiza', 'micra', '1er',
       'yaris', 'colt', '6_reihe', '5_reihe', 'corolla', 'ka', 'tigra',
       'punto', 'vito', 'cordoba', 'galaxy', '100', 'octavia', 'm_klasse',
       'lupo', 'fiesta', 'superb', 'meriva', 'c_max', 'laguna', 'touran',
       '1_reihe', 'm_reihe', 'touareg', 'seicento', 'avensis', 'vivaro',
       'x_reihe', 'ducato', 'carnival', 'boxster', 'signum', 'sharan',
       'zafira', 'rav', 'a5', 'beetle', 'c_reihe', 'phaeton', 'i_reihe',
       'sl', 'insignia', 'up', 'civic', '80', 'mx_reihe', 'omega',
       'sorento', 'z_reihe', 'berlingo', 'clk', 's_max', 'kalos',
 

In [63]:
brand_models={}

for brand in autos["brand"].unique():
    only_brand=autos.loc[autos["brand"]==brand,"model"]
    brand_models[brand]=only_brand.value_counts().head(1).index.tolist()
    
brand_models

{'alfa_romeo': ['156'],
 'audi': ['a4'],
 'bmw': ['3er'],
 'chevrolet': ['andere'],
 'chrysler': ['andere'],
 'citroen': ['andere'],
 'dacia': ['sandero'],
 'daewoo': ['matiz'],
 'daihatsu': ['cuore'],
 'fiat': ['punto'],
 'ford': ['focus'],
 'honda': ['civic'],
 'hyundai': ['i_reihe'],
 'jaguar': ['andere'],
 'jeep': ['grand'],
 'kia': ['andere'],
 'lada': ['niva'],
 'lancia': ['ypsilon'],
 'land_rover': ['freelander'],
 'mazda': ['3_reihe'],
 'mercedes_benz': ['c_klasse'],
 'mini': ['cooper'],
 'mitsubishi': ['colt'],
 'nissan': ['micra'],
 'opel': ['corsa'],
 'peugeot': ['2_reihe'],
 'porsche': ['911'],
 'renault': ['twingo'],
 'rover': ['andere'],
 'saab': ['andere'],
 'seat': ['ibiza'],
 'skoda': ['octavia'],
 'smart': ['fortwo'],
 'sonstige_autos': [],
 'subaru': ['legacy'],
 'suzuki': ['andere'],
 'toyota': ['yaris'],
 'trabant': ['601'],
 'volkswagen': ['golf'],
 'volvo': ['v70']}

Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.

In [64]:
autos["odometer_km"].describe()

count     46473.000000
mean     125688.787038
std       39687.056387
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [65]:
autos["odometer_km"].value_counts(normalize=True)

150000    0.644761
125000    0.104276
100000    0.044069
90000     0.035978
80000     0.029566
70000     0.025499
60000     0.024229
50000     0.021346
40000     0.017150
30000     0.016311
20000     0.015945
5000      0.015773
10000     0.005100
Name: odometer_km, dtype: float64

I choose to split in five groups. Up to 30,000km, up to 60,000km and so forth, till 150,000km

In [66]:
mileage_groups=[30000,60000,90000,150000]
avg_price_m_group={}

for group in mileage_groups:
    price_only_mileage=autos.loc[autos["odometer_km"].between(group-30000,group),"price"]
    avg_price_m_group[group]=price_only_mileage.mean()
    
mileage_price=pd.DataFrame.from_dict(avg_price_m_group,orient="index")
mileage_price.columns=["avg_price"]
mileage_price

avg_price
30000   15483.653706
60000   14337.491968
90000   10168.097256
150000   4123.499569

There is a correspondance between mileage and the average price.
The biggest drops start at over 60,000 kilometers (about 2/3 of the price of cars with 30,000 kilometers or less). Cars with 150,000 kilometers or more are especially cheap (about 1/4).

How much cheaper are cars with damage than their non-damaged counterparts?

In [92]:
#I start off with calculating the means for every model, damaged and non-damages
damaged_non_damaged=pd.DataFrame(columns=["avg_price_undamaged","avg_price_damaged"])
d_nd_values=[]


for model in autos["model"].unique():
    only_model=autos.loc[autos["model"]==model]
    d_nd_values.append(only_model.loc[autos["unrepaired_damage"]=="nein","price"].mean())
    d_nd_values.append(only_model.loc[autos["unrepaired_damage"]=="ja","price"].mean())
    damaged_non_damaged.loc[model]=d_nd_values
    d_nd_values.clear()

    
#No i have to calculate difference for each model and calculate a mean between all differences
mean_difference_d_nd=(damaged_non_damaged["avg_price_undamaged"]-damaged_non_damaged["avg_price_damaged"]).mean()
print(damaged_non_damaged)
print(mean_difference_d_nd)

                    avg_price_undamaged  avg_price_damaged
andere                      8416.156060        2412.225974
7er                        10414.252747        5070.846154
golf                        6099.325103        1856.199346
fortwo                      4011.649533        1507.102564
focus                       4094.413238        1634.811111
voyager                     1957.153846        1477.555556
arosa                       1156.622951         621.357143
megane                      4170.253275        1250.594595
NaN                                 NaN                NaN
a3                          9263.882353        2769.447761
clio                        2807.957704        1088.125000
vectra                      2234.408408         912.227848
scirocco                   12298.285714                NaN
3er                         6961.522899        2336.328638
a4                          8193.432609        2943.366337
911                        74240.922481       20833.3333

The average difference between a damaged use car and a non-damaged used car (between the same car models) is about 3807 Euro. Of course this is just a rough estimation, other variables have to be controlled to determine their influence (e.g. the mileage, etc.).